libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,246 @@
+{
+	if (auto node = std::get_if<FunctionCallGraphBuilder::SpecialNode>(&_node))
+		return *node != FunctionCallGraphBuilder::SpecialNode::Unset;
+
+	return true;
+}
+
+}
+
+namespace solidity::frontend
+{
+
+
+shared_ptr<FunctionCallGraphBuilder::ContractCallGraph> const FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	m_contract = &_contract;
+
+	m_graph = make_shared<ContractCallGraph>(_contract);
+
+	// Create graph for constructor, state vars, etc
+	m_currentNode = SpecialNode::CreationRoot;
+	m_currentDispatch = SpecialNode::CreationDispatch;
+	visitConstructor(
+		_contract,
+		_contract.annotation().linearizedBaseContracts.cbegin() + 1,
+		_contract.annotation().linearizedBaseContracts.cend()
+	);
+	m_currentNode = SpecialNode::Unset;
+	m_currentDispatch = SpecialNode::RuntimeDispatch;
+
+	// Create graph for all publicly reachable functions
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+	{
+		(void)hash;
+		if (auto const* funcDef = dynamic_cast<FunctionDefinition const*>(&functionType->declaration()))
+			if (!m_graph->edges.count(funcDef))
+				visitCallable(funcDef);
+	}
+
+	// Add all CreationDispatch calls to the RuntimeDispatch as well
+	for (auto node: m_graph->edges[SpecialNode::CreationDispatch])
+		add(SpecialNode::RuntimeDispatch, node);
+
+	// Add all external functions to the RuntimeDispatch
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+	{
+		(void)hash;
+		add(SpecialNode::RuntimeDispatch, &functionType->declaration());
+	}
+
+	if (_contract.fallbackFunction())
+		add(SpecialNode::RuntimeDispatch, _contract.fallbackFunction());
+
+	if (_contract.receiveFunction())
+		add(SpecialNode::RuntimeDispatch, _contract.receiveFunction());
+
+	m_contract = nullptr;
+	solAssert(!nodeSet(m_currentNode), "Current node not properly reset.");
+
+	return m_graph;
+}
+
+bool FunctionCallGraphBuilder::visit(Identifier const& _identifier)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	{
+		solAssert(*_identifier.annotation().requiredLookup == VirtualLookup::Virtual, "");
+
+		auto funType = dynamic_cast<FunctionType const*>(_identifier.annotation().type);
+
+		// For events kind() == Event
+		if (funType && funType->kind() == FunctionType::Kind::Internal)
+		{
+			processFunction(callable->resolveVirtual(*m_contract), _identifier.annotation());
+
+			solAssert(nodeSet(m_currentNode), "");
+		}
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(NewExpression const& _newExpression)
+{
+	if (ContractType const* contractType = dynamic_cast<ContractType const*>(_newExpression.typeName().annotation().type))
+		m_graph->createdContracts.emplace(&contractType->contractDefinition());
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(MemberAccess const& _memberAccess)
+{
+	// Bound functions
+	if (auto funType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type))
```

Comment:
Do you think it is possible to make this function much much simpler? Would it work to only consider the function kind and requiredLookup?